presto-spark-base/src/main/java/com/facebook/presto/spark/execution/operator/NativeExecutionOperator.java

```diff
@@ -112,17 +134,86 @@ public void addInput(Page page)
     }
 
     /**
-     * TODO: The getOutput method will call NativeExecutionTask APIs to:
-     *  1. Submit the plan to the external process
-     *  2. Call pollResult method to get latest buffered result.
-     *  3. Call getTaskInfo method to get the TaskInfo and propagate it
-     *  4. Deserialize the polled {@link SerializedPage} to {@link Page} and return it back
+     * The overall workflow of the getOutput method is:
+     * 1. Submit the plan to the external process
+     * 2. Call pollResult method to get latest buffered result.
+     * 3. Call getTaskInfo method to get the TaskInfo and propagate it
+     * 4. Deserialize the polled {@link SerializedPage} to {@link Page} and return it back
      */
     @Override
     public Page getOutput()
     {
-        finished = true;
-        return null;
+        if (finished) {
+            return null;
+        }
+
+        if (process == null) {
+            createProcess();
+            checkState(process != null, "process is null");
+            createTask();
+            checkState(task != null, "task is null");
+            taskStatusFuture = task.start();
+        }
+
+        try {
+            if (taskStatusFuture.isDone()) {
+                // Will throw exception if the  taskStatusFuture is done with error.
+                taskStatusFuture.get();
+                Optional<TaskInfo> taskInfo = task.getTaskInfo();
+                taskInfo.ifPresent(info -> info.getTaskStatus().getFailures().forEach(e -> log.error(e.toException())));
+                Optional<SerializedPage> page = task.pollResult();
+                if (page.isPresent()) {
+                    return processResult(page.get());
+                }
+                else {
+                    finished = true;
+                    return null;
+                }
+            }
+
+            Optional<SerializedPage> page = task.pollResult();
+            return page.map(this::processResult).orElse(null);
```

Comment:
What is the semantics of this method ?

looks like it is returning null, when the output is complete and also when serializedPage is not available. How can the caller stop iterating

Reply:
Return null means currently there is no output - it could be two reasons: 1. C++ side is still processing and hasn't return back  (that what happened at this line here) 2. The C++ side has finished - `taskStatusFuture.isDone() == true` and all data has been processed - in this situation we set the `finished` to true. 

Only the `finished == true` (the caller will call the isFinished() method to get this status) the caller will stop the iterating - otherwise the `getOutput()` method will be continuously called